### PR TITLE
Auto-sync current directory in WinCompat remote session

### PR DESCRIPTION
**Pull request by anmenaga**
# PR Summary
WinCompat uses background Windows PS remote session which has its own current directory. It has to be synchronized with local PS Core current directory for cases when proxy cmdlets rely on current directory.
This PR enables auto-sync of current directory in WinCompat remote session.

This came up in #11411 .

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 1]
```diff
@@ -4848,6 +4853,21 @@ internal void CleanupWindowsPowerShellCompatResources()
             }
         }
 
+        internal static void SyncCurrentLocationHandler(object sender, LocationChangedEventArgs args)
+        {
+            PSSession compatSession = GetWindowsPowerShellCompatRemotingSession();
+            if (compatSession?.Runspace.RunspaceStateInfo.State == RunspaceState.Opened)
```
Comment (anmenaga) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: @PaulHigin Is this the right way to check if remote connection is broken ?

Comment (PaulHigin) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Yes, this is correct.  If the session is not in the 'Opened' state then it cannot run commands.

File: test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1 [hunk 2]
```diff
@@ -327,6 +327,26 @@ Describe "Import-Module from CompatiblePSEditions-checked paths" -Tag "CI" {
...
+            $s = Get-PSSession -Name WinPSCompatSession
+            (Invoke-Command -Session $s {Get-Location}).Path | Should -BeExactly $PWD.Path
+
+            # after local $PWD changes remote $PWD should be synchronized
+            Set-Location -Path ..
+            (Invoke-Command -Session $s {Get-Location}).Path | Should -BeExactly $PWD.Path
+
+            # after WinCompat cleanup local $PWD changes should not cause errors
+            Remove-module $ModuleName -Force
+            Pop-Location
```
Comment (PaulHigin) [inline test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1]: The Pop-Location should be in a try/finally to ensure we revert to the original setting.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs [hunk 3]
```diff
@@ -1978,6 +1978,22 @@ internal override IList<PSModuleInfo> ImportModulesUsingWinCompat(IEnumerable<st
...
+                // make sure that we add registration only once to a multicast delegate
+                SyncCurrentLocationDelegate ??= SyncCurrentLocationHandler;
+                var alreadyregistered = this.SessionState.InvokeCommand.LocationChangedAction?.GetInvocationList().Contains(SyncCurrentLocationDelegate);
+
+                if (!alreadyregistered ?? true)
+                {
+                    this.SessionState.InvokeCommand.LocationChangedAction += SyncCurrentLocationDelegate;
+
+                    // first sync has to be triggered manually
+                    SyncCurrentLocationHandler(this, new LocationChangedEventArgs(null, null, this.SessionState.Path.CurrentLocation));
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs]: use named arguments.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 4]
```diff
@@ -4848,6 +4853,21 @@ internal void CleanupWindowsPowerShellCompatResources()
...
+            {
+                using var ps = System.Management.Automation.PowerShell.Create(RunspaceMode.CurrentRunspace);
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: I do not think you need S.M.A here. 
```suggestion
                using var ps = PowerShell.Create(RunspaceMode.CurrentRunspace);
```

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs (hunk 1 above)
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Is it ok to do nothing when either compatSession is null or RunspaceState is not opened?

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1 (hunk 2 above)
Comment (anmenaga) [reply test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1]: Updated.

File: src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs (hunk 3 above)
Comment (anmenaga) [reply src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs]: Updated.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs (hunk 4 above)
Comment (anmenaga) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Getting this without it:
` The type or namespace name 'Create' does not exist in the namespace 'Microsoft.PowerShell' (are you missing an assembly reference?)`

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs (hunk 1 above)
Comment (anmenaga) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Good observation. That is intentional. If there are problems with the compatSession - they will show up anyway during any WinCompat operation. This handler is called on every location change (which often not related to WinCompat at all) and I don't think it is a good UX if any local directory change would show an error about a broken compatSession.

File: test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1 [hunk 5]
```diff
@@ -345,7 +346,14 @@ Describe "Import-Module from CompatiblePSEditions-checked paths" -Tag "CI" {
 
             # after WinCompat cleanup local $PWD changes should not cause errors
             Remove-module $ModuleName -Force
-            Pop-Location
+            try
```
Comment (PaulHigin) [inline test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1]: Actually, I think all code after the first 'Push-Location' should be in the try.  If any exception is thrown during the test the original location will be reverted through the finally 'Set-Location'.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1 (hunk 5 above)
Comment (anmenaga) [reply test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1]: Updated.

**Review by PaulHigin — APPROVED** [review]
LGTM

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**